File: codifide/parser/tokens.py
"""Keyword table for Codifide surface syntax.

Each keyword has an ASCII form (the primary spelling while we bootstrap on a
US keyboard) and an optional unicode glyph (the canonical display form). The
parser accepts either.
"""
from __future__ import annotations

from typing import Dict

# Keyword -> canonical name used by the parser internally.
KEYWORDS: Dict[str, str] = {
    # ASCII
    "def":     "def",
    "intent":  "intent",
    "sig":     "sig",
    "effects": "effects",
    "pre":     "pre",
    "post":    "post",
    "cand":    "cand",
    "when":    "when",
    "believe": "believe",
    "else":    "else",
    "bottom":  "bottom",
    # Unicode glyphs
    "≡":       "def",
    "⟡":       "intent",
    "σ":       "sig",
    "⚡":       "effects",
    "⊢":       "pre",
    "⊣":       "post",
    "ƒ":       "cand",
    "¿":       "when",
    "⊨":       "believe",
    "⊥":       "bottom",
}

# Operator spellings.
OPERATORS: Dict[str, str] = {
    "<-": "bind",
    "←":  "bind",
    "=>": "arm",
    "⇒":  "arm",
    "++": "concat",
    "⊕":  "concat",
    "->": "arrow",
    "→":  "arrow",
}

File: crates/codifide-interpreter/src/parser/tokens.rs
//! Keyword and operator tables for Codifide surface syntax.
//! Mirrors `codifide/parser/tokens.py`.

/// Map surface keyword (ASCII or Unicode glyph) to canonical name.
pub fn keyword_canon(s: &str) -> Option<&'static str> {
    match s {
        "def" | "≡"      => Some("def"),
        "intent" | "⟡"   => Some("intent"),
        "sig" | "σ"      => Some("sig"),
        "effects" | "⚡"  => Some("effects"),
        "pre" | "⊢"      => Some("pre"),
        "post" | "⊣"     => Some("post"),
        "cand" | "ƒ"     => Some("cand"),
        "when" | "¿"     => Some("when"),
        "cost"           => Some("cost"),
        "believe" | "⊨"  => Some("believe"),
        "else"           => Some("else"),
        "bottom" | "⊥"   => Some("bottom"),
        _ => None,
    }
}

/// Operator spellings (longest first for greedy matching).
pub const OPERATORS: &[(&str, &str)] = &[
    ("<-",  "bind"),
    ("←",   "bind"),
    ("=>",  "arm"),
    ("⇒",   "arm"),
    ("++",  "concat"),
    ("⊕",   "concat"),
    ("->",  "arrow"),
    ("→",   "arrow"),
];

/// Infix comparison/logical operators that desugar to primitive calls.
/// Ordered longest-first so `<=` matches before `<`.
pub const INFIX_OPS: &[(&str, &str)] = &[
    ("<=", "le"),
    (">=", "ge"),
    ("==", "eq"),
    ("!=", "ne"),
    ("<",  "lt"),
    (">",  "gt"),
    ("and","and"),
    ("or", "or"),
];
